torchqdynamics/utils/tensor_types.py

```diff
@@ -67,6 +66,57 @@ def to_tensor(
         )
 
 
+def to_tdtensor(
+    x: TDOperatorLike,
+    dtype: torch.dtype | None = None,
+    device: torch.device | None = None,
+    is_complex: bool = False,
+):
+    """Convert a `TDOperatorLike` object to a `TDTensor` object."""
+    if is_complex:
+        dtype = cdtype(dtype)
+
+    if isinstance(x, get_args(OperatorLike)):
+        x = to_tensor(x, dtype=dtype, device=device, is_complex=True)
+    elif isinstance(x, callable):
+        # check number of arguments and compute at initial time
+        try:
+            x0 = x(0.0)
+        except TypeError:
+            raise TypeError(
+                'Time-dependent operators in the `callable` format should only accept a'
+                ' single argument for time `t`.'
+            )
+
+        # check type, dtype and device match
+        if not isinstance(x0, Tensor):
+            raise TypeError(
+                'Time-dependent operators in the `callable` format should always'
+                ' return a `torch.Tensor` with the same dtype and device as provided'
+                ' to the solver. This avoids type conversion or device transfer at'
+                ' every time step that would slow down the solver. The provided'
+                f' operator is currently of type {type(x0)} instead of {Tensor}.'
+            )
+        elif x0.dtype != dtype:
+            raise TypeError(
+                'Time-dependent operators in the `callable` format should always'
+                ' return a `torch.Tensor` with the same dtype and device as provided'
+                ' to the solver. This avoids type conversion or device transfer at'
+                ' every time step that would slow down the solver. The provided'
+                f' operator is currently of dtype {x0.dtype} instead of {dtype}.'
+            )
+        elif x0.device != device:
+            raise TypeError(
+                'Time-dependent operators in the `callable` format should always'
+                ' return a `torch.Tensor` with the same dtype and device as provided'
+                ' to the solver. This avoids type conversion or device transfer at'
+                ' every time step that would slow down the solver. The provided'
+                f' operator is currently on device {x0.device} instead of {device}.'
+            )
```

Comment:
Let's factor this
```suggestion
        prefix = 'Time-dependent operators in the `callable` format should always'
                ' return a `torch.Tensor` with the same dtype and device as provided'
                ' to the solver. This avoids type conversion or device transfer at'
                ' every time step that would slow down the solver.'
                
        
        if not isinstance(x0, Tensor):
            raise TypeError(f'{prefix} The provided operator is currently of type {type(x0)} instead of {Tensor}.')
        elif x0.dtype != dtype:
            raise TypeError(f'{prefix} The provided operator is currently of dtype {x0.dtype} instead of {dtype}.')
        elif x0.device != device:
            raise TypeError(f'{prefix} The provided operator is currently on device {x0.device} instead of {device}.')
```